### ISSUE TITLE
Protocol abstract async generator method is incompatible with implementation

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#3367]
**Describe the bug**
An abstract async generator method which does not `yield` in its body is reported to be incompatible with its implementation:

```
Method "afoo" overrides class "FooProto" in an incompatible manner
  Return type mismatch: base method returns type "Coroutine[Any, Any, AsyncIterator[str]]", override returns type "AsyncIterator[str]"
    "AsyncIterator[str]" is incompatible with "Coroutine[Any, Any, AsyncIterator[str]]"
```

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That no error would be produced.

**Screenshots or Code**
```python
# pyright: strict

from abc import abstractmethod
from typing import AsyncIterator, Iterator, Protocol
from typing_extensions import reveal_type


class FooProto(Protocol):
    def foo(self) -> Iterator[str]:
        ...

    async def afoo(self) -> AsyncIterator[str]:
        ...

    @abstractmethod
    async def abar(self) -> AsyncIterator[str]:
        ...

    @abstractmethod
    async def abaz(self) -> AsyncIterator[str]:
        return
        yield 


class Foo(FooProto):
    def foo(self) -> Iterator[str]:  # Synchronous iterator is ok
        yield 'foo'

    async def afoo(self) -> AsyncIterator[str]:  # Error
        yield 'afoo'

    async def abar(self) -> AsyncIterator[str]:  # Error
        yield 'abar'

    async def abaz(self) -> AsyncIterator[str]:  # Ok
        yield 'abaz'

```

**VS Code extension or command-line**
Either, 1.1.239

**Comment by erictraut** [comment]
This behavior is intended. The fact that the subclass includes a ` yield` statement means that its actual return type is not an `AsyncIterator[str]` but is instead transformed into a `Coroutine[Any, Any, AsyncIterator[str]]`. Mypy generates the same error in this case.

You can work around it by adding a `yield` statement to the abstract methods in the base class or by explicitly indicating that the return type of the abstract method is `Coroutine[Any, Any, AsyncIterator[str]]`.

**Comment by layday** [comment]
Thanks for the explanation. The simplest 'fix' appears to be to drop the `async` keyword from the abstract method so that it's not interpreted as a coroutine (which it isn't meant to be). It seems unfortunate that the function definition of two disparate types is the same:

```py
async def foo() -> AsyncIterator[str]:  # Function returning an async generator
    yield ''

async def bar() -> AsyncIterator[str]:  # Function returning a coroutine returning an async generator
    return foo()

async def baz():
    async for _ in foo():
        ...
    async for _ in await bar():
        ...
```
